### PR TITLE
Support sharding SQL hint.

### PR DESCRIPTION
**Pull request by dongzl**
ref #20736 .

Changes proposed in this pull request:
- support sharding sql hint.
- add test case.

### CODE REVIEW

**Review by tuichenchuxin — COMMENTED** [review]
I think SQL hint is same as HintManager, the only difference between them is just setting or getting the value way is different.

**Review by tuichenchuxin — COMMENTED** [review]
LGTM. By the way, why databaseValue is only int? It should be comparable is ok.

**Comment by dongzl** [comment]
> LGTM. By the way, why databaseValue is only int? It should be comparable is ok.

OK, fixes it.

**Review by tuichenchuxin — APPROVED** [review]

File: features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/route/engine/type/standard/ShardingStandardRoutingEngine.java [hunk 1]
```diff
@@ -94,6 +98,17 @@ private boolean isRoutingByHint(final ShardingRule shardingRule, final TableRule
                 && shardingRule.getTableShardingStrategyConfiguration(tableRule) instanceof HintShardingStrategyConfiguration;
     }
     
+    private boolean isRoutingBySQLHint() {
+        boolean result = false;
+        if (sqlStatementContext instanceof CommonSQLStatementContext) {
+            Collection<String> tableNames = sqlStatementContext.getTablesContext().getTableNames();
+            for (String each : tableNames) {
+                result = result || ((CommonSQLStatementContext<?>) sqlStatementContext).containsHintShardingValue(each);
```
Comment (strongduanmu) [inline features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/route/engine/type/standard/ShardingStandardRoutingEngine.java:106]: Can you use guard clause here? Just like:

```java
if (CommonSQLStatementContext<?>) sqlStatementContext).containsHintShardingValue(each) {
return true;
}
```

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Review by strongduanmu — APPROVED** [review]
Good job, merged.

**Pull request merged**